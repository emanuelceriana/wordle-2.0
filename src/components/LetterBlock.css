.letterBlock {
  height: 2em;
  width: 2em;
  font-size: 2em;
  text-transform: capitalize;
  margin: 0.1em;
  border-radius: 10%;
  border: 2px solid white;
  display: flex;
  justify-content: center;
  align-items: center;
  background: #464347;
  cursor: pointer;
  position: relative;
  background-color: transparent;
}

.letterBlock.green {
  background-color: rgba(43, 248, 43, 0.459) !important;
}

.letterBlock.yellow {
  background-color: rgba(255, 217, 0, 0.829) !important;
}

.letterBlock.black {
  background-color: rgba(0, 0, 0, 0.514) !important;
}

.letterBlock.focused {
  border: 3px solid gold;
}

.letterBlock:focus {
  outline: none;
}

.letterBlock .ocurrences {
  position: absolute;
  top: 0em;
  right: 0.3em;
  font-size: 0.4em;
  text-transform: lowercase !important;
}

.disabled {
  cursor: not-allowed;
}
