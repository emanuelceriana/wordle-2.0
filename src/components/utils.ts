export const getAmountOfLetters = (inputValues: string[]) => {
  let amountOfLetters = 0;
  inputValues.forEach((letter: string) => letter !== "" && amountOfLetters++);
  return amountOfLetters;
};

export interface ValidationArray {
  value: number;
  ocurrences: number;
  letter: string;
}

export const validateWord = (
  randomWord: string,
  inputValues: string[]
): ValidationArray[] => {
  const validationArray: ValidationArray[] = [];

  inputValues.forEach((letter: string, idx: number) => {
    const validationLetter = { value: -1, ocurrences: 0, letter: "" };

    if (randomWord.indexOf(letter) >= 0) {
      validationLetter.value = 0;
    }
    if (randomWord[idx] === letter) {
      validationLetter.value = 1;
    }

    validationLetter.ocurrences = randomWord
      .split("")
      .filter((l) => l === letter).length;

    validationLetter.letter = letter;

    validationArray[idx] = validationLetter;
  });

  return validationArray;
};

export const isCorrectWord = (validationArray: ValidationArray[]) => {
  const values: number[] = validationArray.map((vA) => vA.value);
  const uniqueValues: number[] = [...new Set(values)];
  return uniqueValues.length === 1 && uniqueValues[0] === 1;
};
