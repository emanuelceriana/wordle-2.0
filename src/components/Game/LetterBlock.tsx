import cn from "classnames";
import { useCallback, useEffect, useRef } from "react";
import { ValidationArray } from "../../utils";
import styles from "./LetterBlock.module.scss";

interface LetterBlockProps {
  focusable: boolean;
  isFocused: boolean;
  value: string;
  idx: number;
  inputValidation: ValidationArray;
  setFocusedInputIdx: (idx: number) => void;
}

export const LetterBlock = ({
  focusable,
  isFocused,
  value,
  idx,
  inputValidation,
  setFocusedInputIdx,
}: LetterBlockProps) => {
  const blockRef = useRef<HTMLDivElement>(null);

  useEffect(() => {
    focusable && isFocused && blockRef?.current?.focus();
  }, [isFocused, focusable]);

  const handleClick = useCallback(() => {
    focusable && blockRef?.current?.focus();
  }, [focusable, blockRef]);

  const handleFocus = useCallback(() => {
    focusable && setFocusedInputIdx(idx);
  }, [focusable]);

  return (
    <div
      className={cn(
        styles.letterBlock,
        { [styles.disabled]: !focusable },
        { [styles.focused]: isFocused },
        {
          [styles["letterBlock__green"]]: inputValidation?.value === 1,
        },
        {
          [styles["letterBlock__yellow"]]: inputValidation?.value === 0,
        },
        {
          [styles["letterBlock__black"]]: inputValidation?.value === -1,
        }
      )}
      tabIndex={-1}
      ref={blockRef}
      onClick={handleClick}
      onFocus={handleFocus}
    >
      {inputValidation && (
        <span className={styles.ocurrences}>
          x{inputValidation?.ocurrences}
        </span>
      )}
      {value}
    </div>
  );
};
