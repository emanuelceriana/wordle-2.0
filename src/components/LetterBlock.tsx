import * as cn from "classnames";
import { useCallback, useEffect, useRef } from "react";
import "./LetterBlock.css";
import { ValidationArray } from "./utils";

interface LetterBlockProps {
  focusable: boolean;
  isFocused: boolean;
  value: string;
  idx: number;
  inputValidation: ValidationArray;
  setFocusedInputIdx: (idx: number) => void;
}

export const LetterBlock = ({
  focusable,
  isFocused,
  value,
  idx,
  inputValidation,
  setFocusedInputIdx,
}: LetterBlockProps) => {
  const blockRef = useRef<HTMLDivElement>(null);

  useEffect(() => {
    focusable && isFocused && blockRef?.current?.focus();
  }, [isFocused, focusable]);

  const handleClick = useCallback(() => {
    focusable && blockRef?.current?.focus();
  }, [focusable, blockRef]);

  const handleFocus = useCallback(() => {
    focusable && setFocusedInputIdx(idx);
  }, [focusable]);

  return (
    <div
      className={cn(
        "letterBlock",
        { disabled: !focusable },
        { focused: isFocused },
        { green: inputValidation?.value === 1 },
        { yellow: inputValidation?.value === 0 },
        { black: inputValidation?.value === -1 }
      )}
      tabIndex={-1}
      ref={blockRef}
      onClick={handleClick}
      onFocus={handleFocus}
    >
      {inputValidation && (
        <span className="ocurrences">x{inputValidation?.ocurrences}</span>
      )}
      {value}
    </div>
  );
};
