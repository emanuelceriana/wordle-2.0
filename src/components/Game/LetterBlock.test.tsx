import React from "react";
import { render, screen, fireEvent, act } from "@testing-library/react";
import { LetterBlock } from "./LetterBlock";
import { ValidationArray } from "../../utils";

interface LetterBlockProps {
  focusable: boolean;
  isFocused: boolean;
  value: string;
  idx: number;
  inputValidation: ValidationArray;
  setFocusedInputIdx: (idx: number) => void;
}

describe("<LetterBlock />", () => {
  const letterBlockProps: LetterBlockProps = {
    focusable: true,
    isFocused: false,
    value: "a",
    idx: 0,
    inputValidation: { letter: "a", ocurrences: 0, value: 1 },
    setFocusedInputIdx: jest.fn(),
  };

  beforeEach(() => {
    jest.clearAllMocks();
    letterBlockProps.focusable = true;
    letterBlockProps.inputValidation = { letter: "a", ocurrences: 0, value: 1 };
  });

  it("should render without errors", () => {
    const { asFragment } = render(<LetterBlock {...letterBlockProps} />);
    expect(asFragment()).toMatchSnapshot();
  });
  it("should be focusable if isFocusable is true", () => {
    render(<LetterBlock {...letterBlockProps} />);

    const letterBlock = screen.getByTestId("letterBlock");

    act(() => {
      fireEvent.click(letterBlock);
    });

    expect(document.activeElement).toStrictEqual(letterBlock);
  });
  it("should not be focusable if isFocusable is false", () => {
    letterBlockProps.focusable = false;

    render(<LetterBlock {...letterBlockProps} />);

    const letterBlock = screen.getByTestId("letterBlock");

    act(() => {
      fireEvent.click(letterBlock);
    });

    expect(document.activeElement).not.toStrictEqual(letterBlock);
  });
  it("should call setFocusedInputIdx when gain focus", () => {
    render(<LetterBlock {...letterBlockProps} />);

    const letterBlock = screen.getByTestId("letterBlock");

    act(() => {
      fireEvent.click(letterBlock);
    });

    expect(letterBlockProps.setFocusedInputIdx).toHaveBeenCalledTimes(1);
  });
  it("should be green if the inputValidation has value 1 (correct letter, correct position)", () => {
    letterBlockProps.value = "e";
    letterBlockProps.inputValidation = { letter: "e", ocurrences: 0, value: 1 };
    render(<LetterBlock {...letterBlockProps} />);

    const letterBlock = screen.getByTestId("letterBlock");

    expect(letterBlock.classList.contains("letterBlock__green")).toBe(true);
  });
  it("should be yellow if the inputValidation has value 0 (only correct letter)", () => {
    letterBlockProps.value = "d";
    letterBlockProps.inputValidation = { letter: "d", ocurrences: 0, value: 0 };
    render(<LetterBlock {...letterBlockProps} />);

    const letterBlock = screen.getByTestId("letterBlock");

    expect(letterBlock.classList.contains("letterBlock__yellow")).toBe(true);
  });
  it("should be black if the inputValidation has value -1 (the random word not contain the letter)", () => {
    letterBlockProps.value = "x";
    letterBlockProps.inputValidation = {
      letter: "x",
      ocurrences: 0,
      value: -1,
    };
    render(<LetterBlock {...letterBlockProps} />);

    const letterBlock = screen.getByTestId("letterBlock");

    expect(letterBlock.classList.contains("letterBlock__black")).toBe(true);
  });
  it("should render the value correctly", () => {
    letterBlockProps.value = "H";

    const { getByText } = render(<LetterBlock {...letterBlockProps} />);

    expect(getByText(letterBlockProps.value)).toBeInTheDocument();
  });
  it("should render the ocurrences correctly", () => {
    letterBlockProps.inputValidation = { letter: "d", ocurrences: 2, value: 0 };

    const { getByText } = render(<LetterBlock {...letterBlockProps} />);

    expect(
      getByText(`x${letterBlockProps.inputValidation.ocurrences}`)
    ).toBeInTheDocument();
  });
  it("should not render ourrences if inputValidation is undefined", () => {
    letterBlockProps.inputValidation = undefined as unknown as ValidationArray;

    const { queryByTestId } = render(<LetterBlock {...letterBlockProps} />);

    expect(queryByTestId("letterBlockOcurrences")).toBeNull();
  });
  it("should call focus to the ref if isFocused and focusable are true", async () => {
    letterBlockProps.isFocused = true;

    jest.spyOn(React, "useEffect").mockImplementation((effect) => effect());
    const mockFocus = jest.fn();
    jest
      .spyOn(React, "useRef")
      .mockReturnValue({ current: { focus: mockFocus } });

    render(<LetterBlock {...letterBlockProps} />);

    act(() => {
      expect(mockFocus).toHaveBeenCalledTimes(1);
    });
  });
});
