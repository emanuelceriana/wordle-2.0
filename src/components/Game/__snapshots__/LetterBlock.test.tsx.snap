// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<LetterBlock /> should render without errors 1`] = `
<DocumentFragment>
  <div
    class="letterBlock letterBlock__green"
    data-testid="letterBlock"
    tabindex="-1"
  >
    <span
      class="ocurrences"
      data-testid="letterBlockOcurrences"
    >
      x0
    </span>
    a
  </div>
</DocumentFragment>
`;
